index.js: fix when sourceFile is /dev/null on windows

saste opened the pull request:
On windows the /dev/null device is not existing, set original to the null string in that case.

Commit:
index.js: handle /dev/null on windows

On windows the /dev/null device is not existing, treat it as a null
device in this case.

Assume /dev/null generates an empty string in case of source file, and
assume /dev/null is a null sync in case of destination file.

This fixes an use case when the patch is generated e.g. with:
diff -u /dev/null test > add.patch

In case of a null output devices, e.g. with:
diff -u test /dev/null > drop.patch

no output is written (as we are implicitly assuming to write to the
/dev/null file).

davidparsson commented:
Hi, and thanks for your contribution. Are you sure this is the best place to solve this? It appears to me this is the first mention of `/dev/null`, and if that's the case I think this issue should be handled outside of this library. What do you think?

saste commented:
So, I think the `applyPatch` function should just accept a patch and do its work. In the case of linux/macos, the `/dev/null` device is existing, the same when working on windows within a MINGW/Cygwin environment, but outside of that environment the current code will fail.

`/dev/null` comes into play when you need to add a file from the void, the unix trick is to assume that the file is coming out of nothing (`/dev/null`) but this does not exist on Windows (I did some research and it seems there is a null driver ([example](https://answers.microsoft.com/en-us/windows/forum/all/windows-10-device-not-working-devnull/358dd2bb-880a-420c-af6c-133ef79be9c5))  but its existence is not working guaranteed and might require additional setup and even in this case might not work correctly, so this looks to me like the most reliable solution (also, this is only impacting windows so it won't change the behavior in other environments and `/dev/null` on Windows is hardly used for other purposes).

`/dev/null` is in fact extremely useful when you want to apply a patch to *create* or *destroy* a file.

```
stefano@mariano:~/tmp/test1$ diff -u /dev/null foo
--- /dev/null   2023-01-24 23:49:50.632183489 +0100
+++ foo 2023-01-25 00:14:57.855934003 +0100
@@ -0,0 +1,5 @@
+a
+b
+c
+d
+
stefano@mariano:~/tmp/test1$ diff -u foo /dev/null
--- foo 2023-01-25 00:14:57.855934003 +0100
+++ /dev/null   2023-01-24 23:49:50.632183489 +0100
@@ -1,5 +0,0 @@
-a
-b
-c
-d
-
```

It's also possible to create a file with using an empty file:
```
stefano@mariano:~/tmp/test1$ touch empty
stefano@mariano:~/tmp/test1$ diff -u empty foo
--- empty       2023-01-25 00:18:50.637073130 +0100
+++ foo 2023-01-25 00:14:57.855934003 +0100
@@ -0,0 +1,5 @@
+a
+b
+c
+d
+
```

but many tools (e.g. `git format-patch` or `diff -r`) will assume `/dev/null`, so this makes it possible to apply patches generated by those tools also on Windows.

Note: probably something similar should be done when deleting a file (that is when `/dev/null` is the target file).

davidparsson commented:
~All right, so that's the use-case. But my concern remains. It seems to me this is a useful but pretty narrow use-case. Wouldn't it be better to apply this on the calling end instead?~

Edit: please disregard this. It has been a few years since I looked at this code.

davidparsson commented:
Hmm. Maybe I'm changing my mind here. If this is how patch files look in these cases, you've got a good point.

What would we do if the destination file is `/dev/null`? It seems reasonable to attempt to fix both of them at once, if the effort is low. Agreed?

Commit:
package: update to latest diff version 5.1.0

saste commented:
I updated the branch, although now I see this is not covering my use case (I was looking for a drop-in multi-platform replacement for `patch`, but now I see that this cannot be covered by `apply-patch`, since it's not supporting multi-file patches and no prefix).

Anyway this is my take to followup your comment: in case the destination file is `/dev/null`, nothing is done at all, following the logic that `/dev/null` is the sync null device. In case of git patches, the logic is a bit different, when you see:
```
diff --git a/test.txt b/test.txt
deleted file mode 100644
index 323fae0..0000000
--- a/test.txt
+++ /dev/null
@@ -1 +0,0 @@
-foobar
```

this means that the `test.txt` file is removed. But for `apply-patch` the model is different, you have a source file and a dest file, you modify the input and the output is written to the dest file (which might be the same as the source), which somehow mismatches the git/patch model.

(In other words, feel free to discard the patches since I'm not really much convinced they are really useful and might break the current model, but you might want to update the `diff` dependency anyway - when testing I noticed the 2.1.1 version was broken in some cases).

Thanks.

davidparsson commented:
Ok, thanks! I think I'll close this then, but I'd anyone would find it useful in the future I'm open for discussions.

Thanks again, and good luck finding anything that suits your needs.
